packages/core-flows/src/definition/cart/steps/update-cart-promotions.ts

```diff
@@ -0,0 +1,107 @@
+import {
+  LinkModuleUtils,
+  ModuleRegistrationName,
+  Modules,
+} from "@medusajs/modules-sdk"
+import { IPromotionModuleService } from "@medusajs/types"
+import { PromotionActions } from "@medusajs/utils"
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+interface StepInput {
+  id: string
+  promo_codes?: string[]
+  action?:
+    | PromotionActions.ADD
+    | PromotionActions.REMOVE
+    | PromotionActions.REPLACE
+}
+
+export const updateCartPromotionsStepId = "update-cart-promotions"
+export const updateCartPromotionsStep = createStep(
+  updateCartPromotionsStepId,
+  async (data: StepInput, { container }) => {
+    const { promo_codes = [], id, action = PromotionActions.ADD } = data
+
+    const remoteLink = container.resolve(LinkModuleUtils.REMOTE_LINK)
+    const remoteQuery = container.resolve(LinkModuleUtils.REMOTE_QUERY)
+    const promotionService = container.resolve<IPromotionModuleService>(
+      ModuleRegistrationName.PROMOTION
+    )
+
+    const existingCartPromotionLinks = await remoteQuery({
+      cart_promotion: {
+        __args: { cart_id: [id] },
+        fields: ["id", "cart_id", "promotion_id"],
+      },
+    })
+
+    const promotionLinkMap = new Map<string, any>(
+      existingCartPromotionLinks.map((link) => [link.promotion_id, link])
+    )
+
+    const promotions = await promotionService.list(
+      { code: promo_codes },
+      { select: ["id"] }
+    )
+
+    const linksToCreate: any[] = []
+    const linksToDismiss: any[] = []
+
+    for (const promotion of promotions) {
+      const linkObject = {
+        [Modules.CART]: { cart_id: id },
+        [Modules.PROMOTION]: { promotion_id: promotion.id },
+      }
+
+      if (
+        action === PromotionActions.ADD ||
+        action === PromotionActions.REPLACE
+      ) {
+        linksToCreate.push(linkObject)
+      }
+
+      if (action === PromotionActions.REMOVE) {
+        const link = promotionLinkMap.get(promotion.id)
+
+        if (link) {
+          linksToDismiss.push(linkObject)
+        }
+      }
+    }
+
+    if (action === PromotionActions.REPLACE) {
+      for (const link of existingCartPromotionLinks) {
+        linksToDismiss.push({
+          [Modules.CART]: { cart_id: link.cart_id },
+          [Modules.PROMOTION]: { promotion_id: link.promotion_id },
+        })
+      }
+    }
+
+    if (linksToDismiss.length) {
```

Comment:
Let's parallelize these calls, no reason not to IMO

Reply:
You can also pass a list of link defs here instead of calling them one by one.

Reply:
@srindom Can you check the latest commit? I think you might be on an older version

Reply:
You are right sorry about that!